### PR TITLE
Use Matcher Names in the DB

### PR DESCRIPTION
**Pull request by ghickman**
This switches out our ForeignKey relations for string fields with the same Matcher names and is the first in a few changes I'm making to pull Matchers completely out of the database.

Note: this change forced me to pull Detectors out a little earlier than I had planned to. They were tied so tightly into the Matching code and did so little it made sense to drop them now.

The `*_match` tables currently have ~120K rows in them so I wouldn't expect a huge amount of downtime for the migrations.  However I've not done migrations on MySQL in a long time so they could do with an extra-detailed review just in case I've missed something!

### CODE REVIEW

File: treeherder/model/models.py [hunk 1]
```diff
@@ -1478,6 +1478,10 @@ class TextLogErrorMetadata(models.Model):
     class Meta:
         db_table = "text_log_error_metadata"
 
+    def __str__(self):
```
Comment (edmorley) [inline treeherder/model/models.py]: Does this model now need a `@python_2_unicode_compatible`?

Comment (ghickman) [reply treeherder/model/models.py]: Good catch, fixed and rebased.

**Comment by ghickman** [comment]
Deployed to prototype and migrations took ~23s based on the New Relic log output in the deploy logs.

**Comment by edmorley** [comment]
> Deployed to prototype and migrations took ~23s based on the New Relic log output in the deploy logs.

I see the following errors on prototype:
`(1054, "Unknown column 'text_log_error_match.matcher_id' in 'field list'")`
https://rpm.newrelic.com/accounts/677903/applications/7385291/filterable_errors#/show/f25eb933-5f8b-11e8-b521-0242ac110006_0_5382/stack_trace?top_facet=transactionUiName&primary_facet=error.class&barchart=barchart

**Comment by ghickman** [comment]
Oh yea good point, how do I do that?

**Comment by edmorley** [comment]
It's on:
https://dashboard.heroku.com/apps/treeherder-prototype/deploy/github

**Comment by ghickman** [comment]
@edmorley @jgraham – further to messages in chat I've opted to not add the discussed `choices` changes.

Having spent the morning testing out the implementation I found it created a confusing layer of abstraction, removing much of the usefulness of having a string directly on *Match instances.  It forced me to add two lookup tables to translate string -> int and int -> string.  Calls to those lookups ended up being peppered around any use of the *Match models.  While the point about increasing the unique constraints size is still valid, I'd rather tackle that when or if it becomes an issue.

While testing out the choices implementation I also took another look at the validation of matcher names, particularly after our discussion around configuring them in settings, and concluded we shouldn't validate them either.  It makes the tests simpler (a matcher name is just a string for tests that don't care about the matcher used) and avoids the need to use any configuration.  In a subsequent PR I'll drop the settings variable and generate available matchers from `Matcher.__subclasses__`.

I'd still like to get consensus on this change though, any further points before I merge?

**Comment by edmorley** [comment]
Sounds fine to me :-)

**Comment by jgraham** [comment]
Did you try something like

```
class Matcher(enum.IntEnum):
    manual = 0
    precise_test = 1
    elastic_search = 2

    @classmethod
    def from_int(cls, value):
        return cls._value2member_map_(value)

[…]
class TextLogErrorMatch(models.Model):
[…]
    matcher_id = models.PositiveSmallIntegerField(choices=(key, item.value) for key, item in Matcher.members.itervalues())
   
    def matcher(self):
        return Matcher.from_int(self.matcher_id)
```

And then you can just use `match.matcher.name` to get the name out. Setting the value should work with the enum directly since `int(value)` works as expected. If you really really want to avoid writing manual conversions you could write a custom field type that will accept an enum, a string, or an int as the input value, and always return an enum in `to_python`.

**Comment by ghickman** [comment]
I didn't try this exact method but thanks for the idea, it still creates a layer of indirection so I'm going to go with the string option.

**Comment by edmorley** [comment]
There are a few other places elsewhere in the overall Treeherder schema where we embed strings like this in the field itself, so it doesn't seem unreasonable to do so here too, given that the alternatives have been explored and add layers of abstraction. Trying to reduce the cognitive overhead of the current autoclassify implementation is a high priority at the moment, and so this PR seems to be making a reasonable tradeoff to help with another step towards doing so.

**Pull request merged**